### PR TITLE
Add HSTS header

### LINKED ISSUE #89: Add HSTS header
Add HSTS header for all responses per https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security.

`Strict-Transport-Security: max-age=31556952; includeSubDomains; preload`

### PR DESCRIPTION
**Pull request by bradleybeddoes**
Updates Rapid Connect to have a HSTS header for all responses.

Closes #89 
References https://github.com/ausaccessfed/infrastructure/issues/340

### CODE REVIEW

**Comment by bradleybeddoes** [comment]
n.b. It looks like we don't have CI for this old project, something we'll have to put on the list.

```
# bradleybeddoes @ Bradleys-MacBook-Pro in ~/Development/aaf/rapidconnect/rapidconnect on git:feature/89-add-hsts-header x [16:36:32]
$ bundle exec rspec
/Users/bradleybeddoes/.rbenv/versions/2.6.6/lib/ruby/gems/2.6.0/gems/activesupport-4.2.11.3/lib/active_support/core_ext/object/duplicable.rb:111: warning: BigDecimal.new is deprecated; use BigDecimal() method instead.
Run options: include {:focus=>true}

All examples were filtered out; ignoring {:focus=>true}

Randomized with seed 63049

ClaimsSet
  for freshdesk services
    sets the name claim
    has no attributes claim
    sets the external_id claim
    sets the email claim
    sets the o claim
    behaves like a jwt claims set
      sets the jti claim
      sets the iat claim
      sets the nbf claim
      sets the exp claim
      sets the iss claim
      sets the typ claim
      sets the aud claim
      #to_jws
        looks valid
        is verified by the secret
        decodes to the claims
  for research services
    behaves like a jwt claims set
      sets the typ claim
      sets the jti claim
      sets the iat claim
      sets the iss claim
      sets the aud claim
      sets the nbf claim
      sets the exp claim
      #to_jws
        is verified by the secret
        looks valid
        decodes to the claims
    behaves like a research claims set
      sets the https://aaf.edu.au/attributes claim
      sets the sub claim
  for zendesk services
    sets the email claim
    has no attributes claim
    sets the o claim
    sets the name claim
    sets the external_id claim
    behaves like a jwt claims set
      sets the iat claim
      sets the iss claim
      sets the nbf claim
      sets the typ claim
      sets the jti claim
      sets the aud claim
      sets the exp claim
      #to_jws
        is verified by the secret
        looks valid
        decodes to the claims
  for auresearch services
    behaves like a research claims set
      sets the https://aaf.edu.au/attributes claim
      sets the sub claim
    behaves like a jwt claims set
      sets the exp claim
      sets the jti claim
      sets the typ claim
      sets the iss claim
      sets the nbf claim
      sets the iat claim
      sets the aud claim
      #to_jws
        looks valid
        decodes to the claims
        is verified by the secret

RapidConnect
  /jwt
    /authnrequest/research
      behaves like a valid service type
        records the retargeted eptid
        creates a JWT
        creates a session
        records an audit log entry
        for the wrong service type
          is expected to be not found
        for a disabled service
          is expected to be forbidden
        for a nonexistent service
          is expected to be not found
        for a blank identifier
          is expected to be not found
    with no authenticated user
      directs to login with entityID included
      directs to login
      forces the response not to be cached
    /authnrequest/zendesk
      behaves like a valid service type
        creates a JWT
        records an audit log entry
        records the retargeted eptid
        creates a session
        for the wrong service type
          is expected to be not found
        for a disabled service
          is expected to be forbidden
        for a nonexistent service
          is expected to be not found
        for a blank identifier
          is expected to be not found
    /authnrequest/freshdesk
      behaves like a valid service type
        creates a session
        records an audit log entry
        redirects to freshdesk in name+val format they define
        records the retargeted eptid
        for the wrong service type
          is expected to be not found
        for a blank identifier
          is expected to be not found
        for a disabled service
          is expected to be forbidden
        for a nonexistent service
          is expected to be not found
    /authnrequest/auresearch
      behaves like a valid service type
        creates a JWT
        creates a session
        records the retargeted eptid
        records an audit log entry
        for a disabled service
          is expected to be forbidden
        for a nonexistent service
          is expected to be not found
        for a blank identifier
          is expected to be not found
        for the wrong service type
          is expected to be not found
  /logout
    performs logout correctly
  /
    sets HSTS header
    shows welcome erb
  /registration
    directs to login if registration attempted when unauthenticated
    sets HSTS header
    sorts the organisations correctly
    shows the registration screen
    /save
      with an invalid endpoint
        behaves like a failed registration
          is rejected
          does not create a service
      in test
        sends no email
        behaves like a successful registration
          ignores a provided service type
          strips spaces from URIs
          redirects to the completed registration page
          creates the service
          sets the timestamp
      with an invalid audience
        behaves like a failed registration
          is rejected
          does not create a service
      in production
        sends an email
        behaves like a successful registration
          redirects to the completed registration page
          sets the timestamp
          ignores a provided service type
          strips spaces from URIs
          creates the service
      with no organisation
        behaves like a failed registration
          is rejected
          does not create a service
      with an excessively short secret
        behaves like a failed registration
          is rejected
          does not create a service
      when an identifier collides
        behaves like a failed registration
          does not create a service
          is rejected
  /administration
    sets HSTS header
    shows the administration dashboard
    directs to login if administration url requested when unauthenticated
    halts with 403 if administration url requested when authenticated user is not an administrator
    /administrators
      allows administrators to be created
      deletes a current administrator
      lists all current administrators
      provides an error when no such administrator is requested to be deleted
      prevents new administrator if name and mail not supplied
      prevents administrator delete if identifier matches current administrator
      prevents duplicate administrators from being saved
      prevents administrator delete if no identifier
      creates new administrator
      prevents new administrator if no identifier
    /services
      lists all current services
      /update
        updates the service
        updates the service type
        with no organisation
          behaves like a failed update
            is rejected
            does not create a service
        with an invalid endpoint
          behaves like a failed update
            does not create a service
            is rejected
        with an invalid audience
          behaves like a failed update
            is rejected
            does not create a service
        with an invalid identifier
          behaves like a failed update
            is rejected
            does not create a service
        with an excessively short secret
          behaves like a failed update
            does not create a service
            is rejected
      /delete/:identifier
        with an invalid identifier
          is expected to be not found
        with a valid identifier
          redirects to the services list
          deletes the service
      /:identifier
        with an invalid identifier
          is expected to be not found
        with a valid identifier
          shows the creation timestamp
          shows a specific service
          for an auresearch service
            shows the endpoint
          for a research service
            shows the endpoint
          with no creation timestamp
            shows a message when no creation timestamp exists
          for a freshdesk service
            shows the endpoint
          for a zendesk service
            shows the endpoint
      /edit/:identifier
        with an invalid identifier
          is expected to be not found
        with a valid identifier
          shows a specific service
      /toggle/:identifier
        with an unknown service
          is expected to be not found
        with an enabled service
          disables the service
          redirects to the service
        with a disabled service
          redirects to the service
          enables the service
  /developers
    shows developers guide
  /login
    redirects to Shibboleth SP SSO with entityID
    sends a redirect to service unknown if original target not in session
    sends a redirect to the original target and populates subject into session when there is a valid Shibboleth SP response
    forces the response not to be cached
    sets HSTS header
    redirects to Shibboleth SP SSO on login request
    sends a 403 response if Shibboleth SP login response contains no session id
    when the attributes contain utf-8 characters
      forces the encoding to be correct
    core attributes are missing
      missing mail
        halts session, shows user an error
      missing cn
        halts session, shows user an error
      missing edupersonscopedaffiliation
        does not halt session, continues
      missing displayname
        halts session, shows user an error
      missing principal
        halts session, shows user an error
  /export
    /basic
      behaves like export API
        export disabled
          404
        export enabled
          authorize header correctly formed
            invalid secret
              403
          authorize header malformed
            403 if not formed correctly
            403 if not supplied
      valid request
        provides json
        200
    /service/:identifier
      invalid service
        404
      behaves like export API
        export enabled
          authorize header malformed
            403 if not formed correctly
            403 if not supplied
          authorize header correctly formed
            invalid secret
              403
        export disabled
          404
      valid service
        200
        provides json
    /services
      valid request
        200
        provides json
      behaves like export API
        export disabled
          404
        export enabled
          authorize header correctly formed
            invalid secret
              403
          authorize header malformed
            403 if not supplied
            403 if not formed correctly

AttributesClaim
  sets the displayname attribute
  sets the mail attribute
  sets the edupersonprincipalname attribute
  sets the edupersonscopedaffiliation attribute
  sets the auedupersonsharedtoken attribute
  has the same edupersontargetedid attribute if email address changes
  has a different edupersontargetedid for another user
  sets the o attribute
  sets the edupersontargetedid attribute
  sets the givenname attribute
  sets the cn attribute
  sets the surname attribute

RapidConnectService
  #from_json
    sets the attributes map
    sets the attribute accessors
    fails on an invalid attribute
    backward compatibility
      considers legacy data valid
      updates the serialized data
      defaults to "research" type
      sets the attributes map
      sets the attribute accessors
  #to_s
    generates a string
    generates a string when identifier is nil
  #identifier!
    persists the identifier
    returns the existing identifier
    leaves an existing identifier
    returns the identifier
  validations
    is expected to validate that :name cannot be empty/falsy
    is expected to allow :type to be ‹"research"›
    is expected to validate that :registrant_name cannot be empty/falsy
    is expected to allow :endpoint to be ‹"https://example.com"›
    is expected not to allow :endpoint to be ‹"https://a b.x.com"›
    is expected not to allow :audience to be ‹"https://a b.x.com"›
    is expected to validate that :secret cannot be empty/falsy
    is expected to allow :endpoint to be ‹"http://example.com"›
    is expected to allow :audience to be ‹"https://example.com"›
    is expected to allow :type to be ‹"auresearch"›
    is expected to validate that :organisation cannot be empty/falsy
    is expected to validate that the length of :secret is at least 16
    is expected not to allow :audience to be ‹"example.com"›
    is expected not to allow :endpoint to be ‹"example.com"›
    is expected to validate that :endpoint cannot be empty/falsy
    is expected to allow :audience to be ‹"http://example.com"›
    is expected to allow :type to be ‹"zendesk"›
    is expected to validate that :audience cannot be empty/falsy
    is expected to validate that :registrant_mail cannot be empty/falsy
    is expected not to allow :type to be ‹"invalid"›
    is expected to allow :type to be ‹"freshdesk"›
    is expected to validate that :created_at looks like a number as long as it is not nil
  #to_json
    contains the attributes
    creates valid json

RapidConnectCLI
  register
    for a new registration
      enables the new service
      is expected to change `RSpec::ExampleGroups::RapidConnectCLI::Register::ForANewRegistration#service_count` by 1
      sets the created_at timestamp
      sets the attributes
      with a missing url option
        fails to create
      output
        indicates successful creation
      with a missing admin option
        fails to create
      with a missing org option
        fails to create
      with a missing secret option
        fails to create
      with a missing name option
        fails to create
      with a missing mail option
        fails to create
      with a missing type option
        sets the type to "research"
        is expected to change `RSpec::ExampleGroups::RapidConnectCLI::Register::ForANewRegistration::WithAMissingTypeOption#service_count` by 1
      with a missing aud option
        fails to create
    when the service already exists
      updates the attributes
      does not update the created_at timestamp
      with a missing name option
        leaves the existing value intact
      with no change
        indicates no change has been made
      with a missing org option
        leaves the existing value intact
      existing service is disabled
        enables the existing service
      with a missing url option
        leaves the existing value intact
      with a missing aud option
        leaves the existing value intact
      with a missing type option
        leaves the existing value intact
      with a missing admin option
        leaves the existing value intact
      with a missing secret option
        leaves the existing value intact
      output
        indicates successful creation
      with a missing mail option
        leaves the existing value intact

Finished in 1.65 seconds (files took 1.8 seconds to load)
273 examples, 0 failures

Randomized with seed 63049

Coverage report generated for RSpec to /Users/bradleybeddoes/Development/aaf/rapidconnect/rapidconnect/coverage. 471 / 472 LOC (99.79%) covered.
```

**Review by mikelkew — APPROVED** [review]
LGTM, assuming that we have adequate test coverage to ensure gem updates didn't break anything.

**Comment by bradleybeddoes** [comment]
We'll find out when it deploys to test fed! 😀

**Review by rianniello — APPROVED** [review]

**Comment by vladimir-mencl-eresearch** [comment]
Hi @bradleybeddoes ,

I see you specified the `preload` option in the HSTS header.

Preload is generally only supported for the base domain (like `aaf.edu.au`) - see https://hstspreload.org/

If you wanted aaf.edu.au preloaded into browsers, you'd have to specify it in the config for your main website. 

It won't hurt here, but will be ignored...

Cheers,
Vlad

File: rapidconnect/Gemfile.lock [hunk 1]
```diff
@@ -217,4 +215,4 @@ DEPENDENCIES
   unicorn
 
 BUNDLED WITH
-   1.16.0
+   2.1.4
```
Comment (vladimir-mencl-eresearch) [inline rapidconnect/Gemfile.lock:218]: Hi @bradleybeddoes , I just tried rolling this out - and my automation tripped on this (bundler 2.1.4) - was this an intentional jump to bundler 2.x ? If so, should `frdatacollator/Gemfile.lock` keep in sync?
Cheers,
Vlad

Comment (bradleybeddoes) [reply rapidconnect/Gemfile.lock:218]: Yes it was, no specific need but you could if you deemed necessary for your env.

**Pull request merged**